### PR TITLE
MINIFICPP-1163 Implemented.

### PR DESCRIPTION
**Pull request by am-c-p-p**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -543,23 +543,30 @@ if(WIN32)
...
 		list(GET VCRUNTIME_REDIST_VERSIONS 0 VCRUNTIME_REDIST_DIR)
 	endif()
-	file(GLOB VCRUNTIME_X86_MERGEMODULES "${VCRUNTIME_REDIST_DIR}/MergeModules/Microsoft_VC*_CRT_x86.msm")
-	file(GLOB VCRUNTIME_X64_MERGEMODULES "${VCRUNTIME_REDIST_DIR}/MergeModules/Microsoft_VC*_CRT_x64.msm")
-	if (NOT VCRUNTIME_X86_MERGEMODULES OR NOT VCRUNTIME_X64_MERGEMODULES)
-		message(FATAL_ERROR "Could not find the VC Redistributable Merge Modules. Please set VCRUNTIME_X86_MERGEMODULE_PATH and VCRUNTIME_X64_MERGEMODULE_PATH manually!")
-	else()
-		list(GET VCRUNTIME_X86_MERGEMODULES 0 VCRUNTIME_X86_MERGEMODULE_PATH)
-		list(GET VCRUNTIME_X64_MERGEMODULES 0 VCRUNTIME_X64_MERGEMODULE_PATH)
+
```
Comment (bakaid) [inline CMakeLists.txt]: This is extra whitespace.

File: CMakeLists.txt [hunk 2]
```diff
@@ -543,23 +543,30 @@ if(WIN32)
...
+	if (NOT VCRUNTIME_X86_REDIST_CRT_DIR OR NOT VCRUNTIME_X64_REDIST_CRT_DIR)
+		message(FATAL_ERROR "Could not find the VC Redistributable. Please set VCRUNTIME_X86_REDIST_CRT_DIR and VCRUNTIME_X64_REDIST_CRT_DIR manually!")
 	endif()
 
 	if(CMAKE_SIZEOF_VOID_P EQUAL 8)
-		message("Using ${VCRUNTIME_X64_MERGEMODULE_PATH} VC Redistributable Merge Module")
-		configure_file("msi/x64.wsi" "msi/x64.wsi" @ONLY)
-		set(CPACK_WIX_EXTRA_SOURCES "${CMAKE_CURRENT_BINARY_DIR}/msi/x64.wsi")
+		message("Using ${VCRUNTIME_X64_REDIST_DIR} VC Redistributables")
```
Comment (bakaid) [inline CMakeLists.txt]: This is non-existent, `VCRUNTIME_X64_REDIST_CRT_DIR` should be used.

File: CMakeLists.txt [hunk 3]
```diff
@@ -543,23 +543,30 @@ if(WIN32)
...
+		file(COPY "${VCRUNTIME_X64_REDIST_CRT_DIR}/msvcp140.dll" DESTINATION "${CMAKE_CURRENT_BINARY_DIR}/redist")
+		file(COPY "${VCRUNTIME_X64_REDIST_CRT_DIR}/msvcp140_1.dll" DESTINATION "${CMAKE_CURRENT_BINARY_DIR}/redist")
+		file(COPY "${VCRUNTIME_X64_REDIST_CRT_DIR}/msvcp140_2.dll" DESTINATION "${CMAKE_CURRENT_BINARY_DIR}/redist")
+		file(COPY "${VCRUNTIME_X64_REDIST_CRT_DIR}/vccorlib140.dll" DESTINATION "${CMAKE_CURRENT_BINARY_DIR}/redist")
+		file(COPY "${VCRUNTIME_X64_REDIST_CRT_DIR}/vcruntime140.dll" DESTINATION "${CMAKE_CURRENT_BINARY_DIR}/redist")
 	elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
-		message("Using ${VCRUNTIME_X86_MERGEMODULE_PATH} VC Redistributable Merge Module")
-		configure_file("msi/x86.wsi" "msi/x86.wsi" @ONLY)
-		set(CPACK_WIX_EXTRA_SOURCES "${CMAKE_CURRENT_BINARY_DIR}/msi/x86.wsi")
+		message("Using ${VCRUNTIME_X86_REDIST_DIR} VC Redistributables")
```
Comment (bakaid) [inline CMakeLists.txt]: This is non-existent, `VCRUNTIME_X86_REDIST_CRT_DIR` should be used.

File: CMakeLists.txt (hunk 1 above)
Comment (am-c-p-p) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt (hunk 3 above)
Comment (am-c-p-p) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt (hunk 2 above)
Comment (am-c-p-p) [reply CMakeLists.txt]: Fixed.

**Comment by am-c-p-p** [comment]
It should be a case when the later doesn't work. If we have this case, we can restore and make it conditional.

**Comment by am-c-p-p** [comment]
If a requirement to not reboot machine, we don't an option other than to put VS Dlls in the bin directory. 
In https://developercommunity.visualstudio.com/content/problem/111499/how-to-detect-if-visual-c-2017-redistributable-is.html, reply `Daniel Griffing, Visual C++ Libraries` - `If you have absolute version-specific requirements, you should choose one of the other deployment options including static linking and/or app local deployment that enable you to control the exact version of your dependencies are in use.`

**Comment by szaszm** [comment]
Let's collect the arguments on both sides and decide which approach is better!

for config option defaulting to merge modules:
* more choice on the use side
* we transparently benefit from later updates to vcredist
* recommended by microsoft (more than DLLs in app directory)
* one less instance of duplicate DLLs on the target system, less disk space needed

for shipping DLLs without merge module option:
* simpler codebase, we don't have to maintain two kinds of packaging
* no need to reboot the machine after installation
* easier to implement

Any more points for/against each side, or new suggestions? Is there a clear industry practice in this question?

**Comment by am-c-p-p** [comment]
Basically it is a comparison of dynamically vs statically linked exes, since exe with VS DLLs in bin directory is similar to statically linked exe in sense that in both cases the exes don't depend on DLLs in Windows/System32 directory.

**Comment by bakaid** [comment]
@am-c-p-p Until we are sure that the new approach works on all systems that the old one supported, I want to keep the option to easily switch between the two methods and not having to dig it up from git history when required.
When the new method has been widely verified by both us and the users, and it proves superior to the Merge Module one, I am perfectly fine with removing the Merge Module one in favour of it.

**Comment by am-c-p-p** [comment]
@szaszm Which dll is missing on Windows 2008 R2?

**Comment by szaszm** [comment]
> @szaszm Which dll is missing on Windows 2008 R2?

`api-ms-win-crt-runtime-l1-1-0.dll`
https://imgur.com/a/nO4Zzue

**Comment by am-c-p-p** [comment]
Added parameter /M (for installer with merge modules) to win_build_vs.bat.
Created new WixWinLocalRedist.wsi for installer with local redistributables Dlls, ideally it would be better to have only one WixWin.wsi, but couldn't figure out simple way to do it.
Please review.

**Comment by bakaid** [comment]
@am-c-p-p Thank you, I will start builds and tests on the new changes.

File: CMakeLists.txt [hunk 4]
```diff
@@ -543,27 +543,60 @@ if(WIN32)
...
-		set(CPACK_WIX_EXTRA_SOURCES "${CMAKE_CURRENT_BINARY_DIR}/msi/x64.wsi")
-	elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
-		message("Using ${VCRUNTIME_X86_MERGEMODULE_PATH} VC Redistributable Merge Module")
-		configure_file("msi/x86.wsi" "msi/x86.wsi" @ONLY)
-		set(CPACK_WIX_EXTRA_SOURCES "${CMAKE_CURRENT_BINARY_DIR}/msi/x86.wsi")
-	else()
-		message(FATAL_ERROR "Could not determine architecture, CMAKE_SIZEOF_VOID_P is unexpected: ${CMAKE_SIZEOF_VOID_P}")
-	endif()
-	set(CPACK_WIX_TEMPLATE "${CMAKE_CURRENT_SOURCE_DIR}/msi/WixWin.wsi")
+    if (INSTALLER_MERGE_MODULES)
```
[the remark this answers is not in the document]
Comment (am-c-p-p) [reply CMakeLists.txt:554]: Fixed.

File: CMakeLists.txt [hunk 5]
```diff
@@ -53,7 +53,7 @@ option(FORCE_WINDOWS "Instructs the build system to force Windows builds when WI
 option(DISABLE_CURL "Disables libCurl Properties." OFF)
 
 option(USE_GOLD_LINKER "Use Gold Linker" OFF)
-
+option(INSTALLER_MERGE_MODULES "Creates installer with merge modules" OFF)
```
Comment (szaszm) [inline CMakeLists.txt:56]: Currently we build with merge modules. This changes to shipped dlls by default. Do we want that?

If you change the PR to the old behavior, please have a look at win_build_vs.bat as well.

Comment (bakaid) [reply CMakeLists.txt:56]: I am fine with it, I feel that the general direction is we want to go forward with the DLL-bundle version once we are sure that it works everywhere/deprecate support everywhere where it doesn't, but for now we have the ability to use the merge module version with a simple configuration.

**Pull request closed without merging**